compiler/rustc_lint/src/builtin.rs

```diff
@@ -1295,15 +1295,80 @@ impl UnreachablePub {
         vis_span: Span,
         exportable: bool,
     ) {
+        fn opt_parent_module(tcx: TyCtxt<'_>, mut did: LocalDefId) -> Option<LocalDefId> {
+            while let Some(parent) = tcx.opt_local_parent(did) {
+                did = parent;
+                if tcx.def_kind(did) == DefKind::Mod {
+                    return Some(did);
+                }
+            }
+            None
+        }
+
         let mut applicability = Applicability::MachineApplicable;
         if cx.tcx.visibility(def_id).is_public() && !cx.effective_visibilities.is_reachable(def_id)
         {
+            // prefer suggesting `pub(super)` instead of `pub(crate)` when possible
+            let new_vis = if let Some(reachable_vis) =
+                cx.effective_visibilities.effective_vis(def_id).map(|effective_vis| {
+                    effective_vis.at_level(rustc_middle::middle::privacy::Level::Reachable)
+                }) {
+                let rustc_middle::ty::Visibility::Restricted(restricted_id) = reachable_vis else {
+                    unreachable!()
+                };
+
+                if let Some(parent) = opt_parent_module(cx.tcx, def_id.into()) {
+                    if let Some(parent_parent) = opt_parent_module(cx.tcx, parent) {
+                        if &parent_parent == restricted_id {
+                            // current item is restricted to the parent of it's parent (aka super)
+                            "pub(super)"
+                        } else {
+                            // current item is not restricted to the parent of it's parent
+                            "pub(crate)"
+                        }
+                    } else {
+                        // unreachable pub at crate root, is that even possible?
+                        "pub(crate)"
+                    }
+                } else {
+                    // current item has no parent, is that even possible?
+                    "pub(crate)"
+                }
+            } else {
+                // current item can't be reach from crate, probably in an anon-const
+                if let Some(direct_parent) = cx.tcx.opt_local_parent(def_id.into()) {
+                    if matches!(
+                        cx.tcx.def_kind(direct_parent),
+                        DefKind::AnonConst | DefKind::InlineConst
+                    ) {
+                        // current item at top level const-anon
+                        ""
+                    } else if let Some(direct_parent_parent) =
+                        cx.tcx.opt_local_parent(direct_parent)
+                        && matches!(
+                            cx.tcx.def_kind(direct_parent_parent),
+                            DefKind::AnonConst | DefKind::InlineConst
+                        )
+                    {
+                        // current item is restricted to the parent of it's parent (aka super)
```

Comment:
I don't think the logic here makes sense, `super` is specifically a parent module, and not any parent node with a def id.

Reply:
Just keep suggesting `pub(crate)` for items not tracked in the `effective_visibilities` table.
Anything more restrictive may be incorrect (but we don't know because it's not tracked), and it's not an important use case to address precisely.

Reply:
I agree that it's not an important use-case, so I have revert the anon-const handling, but I still think that we should at some point handle it.

> I don't think the logic here makes sense, `super` is specifically a parent module, and not any parent node with a def id.

Yes but one cannot reach inside a anon-const/inline-const, that means that the `super` mod cannot reach inside and therefore neither `pub(super)` nor `pub(crate)` makes sense.

Anyway, that's not important. I've reverted the handling.

Reply:
An example where `S` is not in the table, but `pub(super)` is not enough.

```rust
#![feature(rustc_attrs)]
#![allow(internal_features)]

const _: u8 = {
    pub mod m1 {
        pub mod m2 {
            #[rustc_effective_visibility]
            pub struct S {}
        }
        
        pub use m2::S;
    }
    
    let _ = m1::S {};
    
    10
};
```

Reply:
>but I still think that we should at some point handle it

I think the right way would be to remove optimizations preventing precise reachability tracking from looking into blocks, it just may be somewhat costly.

Reply:
> An example where `S` is not in the table, but `pub(super)` is not enough.

Completely agree, and that's why my code only try to dealt with top and 1-level in anon-const and bail-out for anything else, like your example.

> I think the right way would be to remove optimizations preventing precise reachability tracking from looking into blocks, it just may be somewhat costly.

Good to know. If I find the time, I will experiment with that.